packages/react-compose/src/types.ts

```diff
@@ -1,30 +1,64 @@
 import * as React from 'react';
 
-export type ComposedComponent<
-  InputProps = {},
-  InputStylesProps = {},
-  ParentProps = {},
-  ParentStylesProps = {}
-> = React.FunctionComponent<InputProps & ParentProps> & {
-  fluentComposeConfig: ComposePreparedOptions<InputProps, InputStylesProps, ParentProps, ParentStylesProps>;
+//
+// "as" type safety
+//
+
+export type PropsOfElement<
+  E extends keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>
+> = JSX.LibraryManagedAttributes<E, React.ComponentPropsWithRef<E>>;
+
+export type ComponentWithAs<E extends React.ElementType, P> = (<EE extends React.ElementType = E>(
+  props: Omit<PropsOfElement<EE>, 'as' | keyof P> & { as?: EE } & P,
+) => JSX.Element) & {
+  displayName?: string;
+
+  defaultProps?: Partial<P & { as: E }>;
+  propTypes?: React.WeakValidationMap<P> & {
+    as: React.Requireable<string | ((props: any, context?: any) => any) | (new (props: any, context?: any) => any)>;
+  };
 };
 
+//
+// Compose types
+//
+
+export type ComposedComponent<P = {}> = React.FunctionComponent<P> & {
+  fluentComposeConfig: ComposePreparedOptions;
+};
+
+export type InputComposeComponent<P = {}> = React.FunctionComponent<P> & {
```

Comment:
What does this name mean? Is that "input-only" compose component, like form input elements only? Or the component which compose takes as an input?

Reply:
The component that compose takes as an input... I am open for some suggestions here to be honest ;)